### PR TITLE
Add support for isApproval and hbar transfers

### LINKED ISSUE #4087: Add support for isApproval and hbar transfers in the transferPrecompile class
### Problem

Epic: #4084 

Currently, the isApproval flags is always set to false.  
Also,  there is no support for transferring hbars as part of this precompile.

### Solution

Update the logic to use the passed in parameter for setting the isApproval flag.
Add changes that reflect the transfer of hbars for processing.  
Update the fee charged to include the cost of transferring hbars.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by lukelee-sl**
Signed-off-by: lukelee-sl <luke.lee@swirldslabs.com>

**Description**:
Ensure that the isApproval parameter passed in from the cryptoTransfer function is utilized when performing the transfers.

Add HBAR transfers to the list of changes to perform.

**Related issue(s)**:

Fixes #4087 

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java
Comment (dimitar-dinev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:589]: Maybe extract `EntityIdUtils.accountIdFromEvmAddress(senderAddress)` to a variable before all the loops?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java
Comment (dimitar-dinev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:600]: ⚠️  The hbar transfers loop is currently nested in the token transfers loop, it should be extracted, otherwise hbar balance changes may be repeated multiple times or not even acknowledged if there are no token transfers.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java
Comment (dimitar-dinev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:583]: Not related to your PR, but I'm wondering whether we need to construct a separate 
`final List<BalanceChange> changes = new ArrayList<>()` on each iteration, add the changes in it only to then add them to the `allChanges` list, instead of directly using the `allChanges` list?

**Review by dimitar-dinev — COMMENTED** [review]
Overall LGTM, had only 1 concern when constructing the balance changes. Not blocking the PR to not keep you from merging during your day, but please see the comment.

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/services/store/contracts/precompile/TransferPrecompilesTest.java
Comment (tdermendzhievv) [inline hedera-node/hedera-mono-service/src/test/java/com/hedera/services/store/contracts/precompile/TransferPrecompilesTest.java:1492]: Could you add  also test case for hbar transfer + fungible transfer + non fungible?

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/services/store/contracts/precompile/TransferPrecompilesTest.java
Comment (tdermendzhievv) [inline hedera-node/hedera-mono-service/src/test/java/com/hedera/services/store/contracts/precompile/TransferPrecompilesTest.java:1420]: Could you add test case for multiple hbar transfers (List<SyntheticTxnFactory.HbarTransfer> hbarTransfers > 1)?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:600]: yikes! thanks for catching this.  Had it right in my other repo but clearly this is wrong the way it is.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:583]: This seems like a good optimization.  I can do it while I'm here.

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/services/store/contracts/precompile/TransferPrecompilesTest.java
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/test/java/com/hedera/services/store/contracts/precompile/TransferPrecompilesTest.java:1535]: done

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/services/store/contracts/precompile/TransferPrecompilesTest.java
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/test/java/com/hedera/services/store/contracts/precompile/TransferPrecompilesTest.java:1426]: done

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java
Comment (Nana-EC) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:185]: Please add a comment for this logic

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java
Comment (Nana-EC) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:325]: Q: Is there a  cost capturing the approval transfer.
For example do we capture the work done by the network to verify the approved transfer is permissible?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java
Comment (Nana-EC) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:571]: Q: Should there be an isApproval flag for NTS exchanges here?

**Review by Nana-EC — CHANGES_REQUESTED** [review]
LG, but clarifying questions.

**Review by dimitar-dinev — APPROVED** [review]

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:325]: I do not see this use case called out in the [fee schedule](https://docs.hedera.com/guides/mainnet/fees#transaction-and-query-fees)  Nor do I see any extra fees being charged for approved transfers in HAPI.  There is a fee for setting the approval in the first place.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:571]: The isApproval flag is set inside changingNftOwership [method] https://github.com/hashgraph/hedera-services/blob/8776416c86cdf1f1450ec5fc6cbb961e1f78ddbe/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/ledger/BalanceChange.java#L109

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:188]: done

**Review by Nana-EC — APPROVED** [review]
LG

**Pull request merged**
